### DISCUSSION TITLE
V0.1.3 Released!

### DISCUSSION THREAD

**Discussion by ver217** [hpcaitech/ColossalAI#780] [Announcements]
## Overview
Here are the main improvements of this release:
1. Gemini: Heterogeneous memory space manager
2. Refactor the API of  pipeline parallelism

### Features
* [zero] initialize a stateful tensor manager by @feifeibear in https://github.com/hpcaitech/ColossalAI/pull/614
* [pipeline] refactor pipeline by @YuliangLiu0306 in https://github.com/hpcaitech/ColossalAI/pull/679
* [zero] stateful tensor manager by @ver217 in https://github.com/hpcaitech/ColossalAI/pull/687
* [zero] adapt zero hooks for unsharded module by @1SAA in https://github.com/hpcaitech/ColossalAI/pull/699
* [zero] refactor memstats collector by @ver217 in https://github.com/hpcaitech/ColossalAI/pull/706
* [zero] improve adaptability for not-shard parameters by @1SAA in https://github.com/hpcaitech/ColossalAI/pull/708
* [zero] check whether gradients have inf and nan in gpu by @1SAA in https://github.com/hpcaitech/ColossalAI/pull/712
* [refactor] refactor the memory utils by @feifeibear in https://github.com/hpcaitech/ColossalAI/pull/715
* [util] support detection of number of processes on current node by @FrankLeeeee in https://github.com/hpcaitech/ColossalAI/pull/723
* [utils] add synchronized cuda memory monitor by @1SAA in https://github.com/hpcaitech/ColossalAI/pull/740
* [zero] refactor ShardedParamV2 by @1SAA in https://github.com/hpcaitech/ColossalAI/pull/742
* [zero] add tensor placement policies by @ver217 in https://github.com/hpcaitech/ColossalAI/pull/743
* [zero] use factory pattern for tensor_placement_policy by @feifeibear in https://github.com/hpcaitech/ColossalAI/pull/752
* [zero] refactor memstats_collector by @1SAA in https://github.com/hpcaitech/ColossalAI/pull/746
* [gemini] init genimi individual directory by @feifeibear in https://github.com/hpcaitech/ColossalAI/pull/754
* refactor shard and gather operation by @1SAA in https://github.com/hpcaitech/ColossalAI/pull/773

### Bug Fix
* [zero] fix init bugs in zero context by @1SAA in https://github.com/hpcaitech/ColossalAI/pull/686
* [hotfix] update requirements-test by @ver217 in https://github.com/hpcaitech/ColossalAI/pull/701
* [hotfix] fix a bug in 3d vocab parallel embedding by @kurisusnowdeng in https://github.com/hpcaitech/ColossalAI/pull/707
* [compatibility] fixed tensor parallel compatibility with torch 1.9 by @FrankLeeeee in https://github.com/hpcaitech/ColossalAI/pull/700
* [hotfix]fixed bugs of assigning grad states to non leaf nodes by @Gy-Lu in https://github.com/hpcaitech/ColossalAI/pull/711
* [hotfix] fix stateful tensor manager's cuda model data size by @ver217 in https://github.com/hpcaitech/ColossalAI/pull/710
* [bug] fixed broken test_found_inf by @FrankLeeeee in https://github.com/hpcaitech/ColossalAI/pull/725
* [util] fixed activation checkpointing on torch 1.9 by @FrankLeeeee in https://github.com/hpcaitech/ColossalAI/pull/719
* [util] fixed communication API with PyTorch 1.9 by @FrankLeeeee in https://github.com/hpcaitech/ColossalAI/pull/721
* [bug] removed zero installation requirements by @FrankLeeeee in https://github.com/hpcaitech/ColossalAI/pull/731
* [hotfix] remove duplicated param register to stateful tensor manager by @feifeibear in https://github.com/hpcaitech/ColossalAI/pull/728
* [utils] correct cpu memory used and capacity in the context of multi-process by @feifeibear in https://github.com/hpcaitech/ColossalAI/pull/726
* [bug] fixed grad scaler compatibility with torch 1.8 by @FrankLeeeee in https://github.com/hpcaitech/ColossalAI/pull/735
* [bug] fixed DDP compatibility with torch 1.8 by @FrankLeeeee in https://github.com/hpcaitech/ColossalAI/pull/739
* [hotfix] fix memory leak in backward of sharded model by @ver217 in https://github.com/hpcaitech/ColossalAI/pull/741
* [hotfix] fix initialize about zero by @ver217 in https://github.com/hpcaitech/ColossalAI/pull/748
* [hotfix] fix prepare grads in sharded optim by @ver217 in https://github.com/hpcaitech/ColossalAI/pull/749
* [hotfix] layernorm by @kurisusnowdeng in https://github.com/hpcaitech/ColossalAI/pull/750
* [hotfix] fix auto tensor placement policy by @ver217 in https://github.com/hpcaitech/ColossalAI/pull/753
* [hotfix] fix reuse_fp16_shard of sharded model by @ver217 in https://github.com/hpcaitech/ColossalAI/pull/756
* [hotfix] fix test_stateful_tensor_mgr by @ver217 in https://github.com/hpcaitech/ColossalAI/pull/762
* [compatibility] used backward-compatible API for global process group by @FrankLeeeee in https://github.com/hpcaitech/ColossalAI/pull/758
* [hotfix] fix the ckpt hook bugs when using DDP by @Gy-Lu in https://github.com/hpcaitech/ColossalAI/pull/769
* [hotfix] polish sharded optim docstr and warning by @ver217 in https://github.com/hpcaitech/ColossalAI/pull/770

### Unit Testing
* [ci] replace the ngc docker image with self-built pytorch image by @FrankLeeeee in https://github.com/hpcaitech/ColossalAI/pull/672
* [ci] fixed compatibility workflow by @FrankLeeeee in https://github.com/hpcaitech/ColossalAI/pull/678
* [ci] update workflow trigger condition and support options by @FrankLeeeee in https://github.com/hpcaitech/ColossalAI/pull/691
* [ci] added missing field in workflow by @FrankLeeeee in https://github.com/hpcaitech/ColossalAI/pull/692
* [ci] remove ipc config for rootless docker by @FrankLeeeee in https://github.com/hpcaitech/ColossalAI/pull/694
* [test] added missing decorators to model checkpointing tests by @FrankLeeeee in https://github.com/hpcaitech/ColossalAI/pull/727
* [unitest] add checkpoint for moe zero test by @1SAA in https://github.com/hpcaitech/ColossalAI/pull/729
* [test] added a decorator for address already in use error with backward compatibility by @FrankLeeeee in https://github.com/hpcaitech/ColossalAI/pull/760
* [test] refactored with the new rerun decorator by @FrankLeeeee in https://github.com/hpcaitech/ColossalAI/pull/763

### Documentation
* add PaLM link by @binmakeswell in https://github.com/hpcaitech/ColossalAI/pull/704
* [doc] removed outdated installation command by @FrankLeeeee in https://github.com/hpcaitech/ColossalAI/pull/730
* add video by @binmakeswell in https://github.com/hpcaitech/ColossalAI/pull/732
* [readme] polish readme by @feifeibear in https://github.com/hpcaitech/ColossalAI/pull/764
* [readme] sync CN readme by @binmakeswell in https://github.com/hpcaitech/ColossalAI/pull/766

### Miscellaneous
* [Bot] Synchronize Submodule References by @github-actions in https://github.com/hpcaitech/ColossalAI/pull/556
* [Bot] Synchronize Submodule References by @github-actions in https://github.com/hpcaitech/ColossalAI/pull/695
* [refactor] zero directory  by @feifeibear in https://github.com/hpcaitech/ColossalAI/pull/724
* [Bot] Synchronize Submodule References by @github-actions in https://github.com/hpcaitech/ColossalAI/pull/751

**Full Changelog**: https://github.com/hpcaitech/ColossalAI/compare/v0.1.2...v0.1.3

<hr /><em>This discussion was created from the release <a href='https://github.com/hpcaitech/ColossalAI/releases/tag/v0.1.3'>V0.1.3 Released!</a>.</em>

**Comment by DustinEwan** [comment]
> [zero] check whether gradients have inf and nan in gpu by @1SAA in https://github.com/hpcaitech/ColossalAI/pull/712

I occasionally see a warning from this: `WARNING colossalai - ShardedOptimizerV2 - WARNING: found inf during ShardedOptimV2 step`

Is there something I should be doing about this, or is this just like a "hey it happens, but thought you should know" sort of warning?

**Reply by feifeibear** [reply]
The warning is also existing in the other frameworks using mixed precision training. If you see it occasionally that's fine. The system just skipped the grad overflowing iterations. If you see nan in every iteration, you should tune the hyperparameters, a smaller learning rate for example to avoid grad value exceeding the range of float fp16.
